docs/src/developer/design/user_api.rst

```diff
@@ -36,133 +55,359 @@ functional-style and cmake-based
+on C++; build targets for most other coding languages will have similar needs.
+For a typical C++ target we need to specify the:
+
+- name of the build target,
+- source files defining the build target's implementation,
+- header files defining the build target's public :term:`API`, and
+- build target's dependencies (including other build targets).
+
+The proposed CMaize APIs are:
+
+.. code-block:: CMake
+
+   # Declaring a build target for a (C++) library
+   cmaize_add_library(
+      <name>
+      SOURCE_DIR <where_the_source_files_are_located>
+      INCLUDE_DIRS <directories_containing_header_files>
+      DEPENDS <dependency0> <dependency1> ...
+   )
+
+   # Declaring a build target for a (C++) executable is similar
+   cmaize_add_executable(
+      <name>
+      SOURCE_DIR <where_the_source_files_are_located>
+      INCLUDE_DIRS <directories_containing_header_files>
+      DEPENDS <dependency0> <dependency1>
+   )
+
+Like the "Find Dependencies" step before it, the APIs for defining build targets
+are designed primarily for collecting information pertaining to the build
+target. Unlike the "Find Dependencies" step the backend of API calls for
+defining build targets is CMake. The result of calling these methods are
+properly configured CMake targets.
+
+Test Project
+============
+
+After targets are built the next step is to test that they were built correctly.
+Testing build targets with CMake often requires:
+
+- finding dependencies of the testing framework,
+- building testing targets which consume the project's build targets, and
+- registering the tests with CTest.
+
+It is also worth noting that tests are often built conditionally (e.g., a
+:term:`build system` typically does not build the tests of dependencies built
+during the "Find Dependencies" step). To that end CMake defines the
+``BUILD_TESTING`` variable; when set to ``TRUE`` tests are built, otherwise they
+are not. Proposed CMaize APIs for testing a project:
+
+.. code-block:: CMake
+
+   cmaize_find_or_build_test_dependency(
+       <name>
+       URL <where_on_the_internet_to_download_from>
+       VERSION <the_version_you_want>
+       BUILD_TARGET <target_to_build>
+       FIND_TARGET <target_representing_package>
+       CMAKE_ARGS <configuration_options_to_set>
+   )
+
+   cmaize_add_test_library(
+       <name>
+       SOURCE_DIR <where_the_source_files_are_located>
+       INCLUDE_DIRS <directories_containing_header_files>
+       DEPENDS <dependency0> <dependency1> ...
+   )
+
+   cmaize_add_test_executable(
+       <name>
+       SOURCE_DIR <where_the_source_files_are_located>
+       INCLUDE_DIRS <directories_containing_header_files>
+       DEPENDS <dependency0> <dependency1> ...
+   )
+
+   # This is actually CTest's add_test command
+   add_test(NAME <name> COMMAND)
+
+   # This is a convenience function for the common scenario where the
+   # add_test call simply calls the executable arising from the
+   # cmaize_add_test_executable
+   cmaize_add_test(
```

Comment:
Currently, `cmaize_add_test` is named `cmaize_add_tests` and is a wrapper around calling `cmaize_add_executable` and registering it to CTest with `add_test`.

Reply:
I'm advocating for renaming it to `cmaize_add_test` for consistency. (We can keep `cmaize_add_test` and forward it so people don't need to remember whether it's plural or not).

Reply:
Sounds good, I am fine with either option.